Add container resolver for otk: otk-resolve-containers (COMPOSER-2293)

achilleas-k opened the pull request:
New `cmd` for resolving containers for otk.

The resolver is a thin wrapper around the container.Resolver.  It can handle multiple containers for the same architecture in one call.

supakeen reviewed (commented):
Going to leave the same comment as on #935: what about the `const`. We can resolve both in one go there. Otherwise very nice.

mvo5 commented on cmd/otk-resolve-containers/main_test.go:
```diff
@@ -0,0 +1,194 @@
+	resolver "github.com/osbuild/images/cmd/otk-resolve-containers"
+)
+
+const rootLayer = `H4sIAAAJbogA/+SWUYqDMBCG53lP4V5g9x8dzRX2Bvtc0VIhEIhKe/wSKxgU6ktjC/O9hMzAQDL8
```
Maybe a small comment how this is generated, what it represents etc?

mvo5 replied:
Nitpick (and fine for a followup or fine for me to take if you write a coment here in GH) - but I'm still curious where this comes from.

achilleas-k replied:
Sorry.  Forgot to add this.  I'll include a comment.

mvo5 commented on cmd/otk-resolve-containers/main_test.go:
```diff
@@ -0,0 +1,194 @@
+			require := require.New(t)
+			assert := assert.New(t)
+
+			amd64input := map[string]interface{}{
```
I may sound silly here, but I'm a fan of writing the json out mostly because to me a test is also documentation and having a look into the test file to see what the json input and expected json output looks like feels rather nice. I can see it's tedious and work so feel free to ignore (both there and in the expected output)

achilleas-k replied:
In this case it's incredibly tedious but for the output it's also mostly impossible to get right because the order or elements (and sometimes keys) isn't predictable.  It's often stable, and the elements are stable in our container resolver, but modifying the tests in any way would require adjusting the expected output and figuring out the new order of elements.

Not to mention, the error output if it fails is a lot less readable when it's just comparing json in string format.

achilleas-k replied:
I just noticed I accidentally left the test running with only 1 ref when the comment says 10 (it should be 10).  But I'll add this test with raw json and a single ref too.

mvo5 commented on cmd/otk-resolve-containers/main_test.go:
```diff
@@ -0,0 +1,194 @@
+		errSubstr string
+	}
+
+	testCases := map[string]testCase{
```
I wonder if we could get away with testing slightly less cases here? It's fine to keep as is as it's already written but at this level we are mostly caring about that when we call `Run()` and it results in an error but the details of the various ways that `run` can fail. Which seems mostly `resolver.Finish()` so testing that is great but at the same time just checking one way (or even mocking) is probably enough.

achilleas-k replied:
I felt it was more useful to make sure that the error message provides useful information for a command line tool, vs simply testing that "error exists".
Of course, this all comes straight from `resolver.Finish()` like you mentioned, so maybe here is not the place to test for it.  I do like that we're making sure that _the container resolver command line tool isn't "eating" parts of the underlying error messages_.

mvo5 replied:
Sorry, I probably expressed this not very well, it's great to have a test that ensures we do pass the error properly and not "eat" part of the underlying error message. I was mostly wondering if we need to test this for all the cases. But as I wrote, it's fine as is.

achilleas-k replied:
Not really necessary to cover all cases but I think it's nice to have.

mvo5 commented on cmd/otk-resolve-containers/main_test.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func TestResolver(t *testing.T) {
+	registry, refs := createTestRegistry()
```
Thanks, this looks like a really nice test abstraction and given that `NewResolver()` is hard to mock it's seems the best approach to test this.

achilleas-k replied:
Thank whoever wrote that mock container registry :D

mvo5 reviewed (commented):
Thanks! This looks nice, ideally we would apply the outcome of the `const` discusison here and there are some ideas/suggestions inline for your considerations, mostly about spelling out the inputs/outputs slightly more explicitly

achilleas-k commented:
> Going to leave the same comment as on #935: what about the `const`. We can resolve both in one go there. Otherwise very nice.

Done!  Moved output under `const` and also added an extra level, `containers`, in case we need to add more to the output at some point, because the previous top-level object (the one under `.tree.const`) is an array.

mvo5 commented on cmd/otk-resolve-containers/main.go:
```diff
@@ -0,0 +1,117 @@
+	// The architecture of the image
+	Arch string `json:"arch"`
+
+	LocalStorage bool  `json:"local-storage,omitempty"`
```
This means that "LocalStorage: false" will not be part of the json output, not sure that is desirable or now, we could consider not exposing it at all as I'm not sure how we would use it in an otk file or just remove the omitempty(?)

achilleas-k replied:
I think you're right.  We should probably hide it completely and see if we ever need it.

achilleas-k replied:
Dropped!

mvo5 reviewed (dismissed):
Thank you! This is nice and I really like the detail that pkg/containers/spec.go:Spec is not directly exposed as we have now more control.

There are some testing nitpicks/suggestions inline but as they only affect testing (which is already comprehensive) and are very opinionated (not neccessarily in a good way, sorry for that!)  - the testing is very nice and comprehensive (please don't get me wrong here). *If* you like the ideas I am happy to do a followup, if you don't I will just be silent :)

The only thing that we might consider is the coment/question about `LocalStorage` (but given that most of the times I think it will be false/not visible(?) it should cause no harm even if we merge as is).

supakeen reviewed (dismissed):
Needs the other comments resolved, awesome.

mvo5 reviewed (approved):
Thank you

achilleas-k commented on cmd/otk-resolve-containers/main_test.go:
```diff
@@ -0,0 +1,194 @@
+			require := require.New(t)
+			assert := assert.New(t)
+
+			amd64input := map[string]interface{}{
```
Yup.  Renaming.

achilleas-k replied:
Done
